fs/fs_zenfs.cc

```diff
@@ -1304,7 +1317,122 @@ void ZenFS::GetZenFSSnapshot(ZenFSSnapshot& snapshot,
+    s = MigrateFileExtents(it.first, it.second);
+    if (!s.ok()) break;
+  }
+  return s;
+}
+
+IOStatus ZenFS::MigrateFileExtents(
+    const std::string& fname,
+    const std::vector<ZoneExtentSnapshot*>& migrate_exts) {
+  IOStatus s = IOStatus::OK();
+  Info(logger_, "MigrateFileExtents, fname: %s, extent count: %lu",
+       fname.data(), migrate_exts.size());
+  // The file may be deleted by other threads, better double check.
+  auto zfile = GetFile(fname);
+  if (zfile == nullptr || zfile->IsOpenForWR()) {
+    return IOStatus::OK();
+  }
+
+  std::vector<ZoneExtent*> extents = zfile->GetExtents();
+  for (ZoneExtent* ext : extents) {
+    // Check if current extent need to be migrated
+    auto it = std::find_if(migrate_exts.begin(), migrate_exts.end(),
+                           [&](const ZoneExtentSnapshot* ext_snapshot) {
+                             return ext_snapshot->start == ext->start_ &&
+                                    ext_snapshot->length == ext->length_;
+                           });
+
+    if (it == migrate_exts.end()) {
+      Info(logger_, "Migrate extent not found, ext_start: %lu", ext->start_);
+      continue;
+    }
+
+    Zone* old_zone = ext->zone_;
+    Zone* target_zone = nullptr;
+
+    // Allocate a new migration zone.
+    s = zbd_->TakeMigrateZone(&target_zone, zfile->GetWriteLifeTimeHint(),
+                              ext->length_);
+    if (!s.ok()) {
+      continue;
+    }
+
+    if (target_zone == nullptr) {
+      zbd_->ReleaseMigrateZone(target_zone);
+      Info(logger_, "Migrate Zone Acquire Failed, Ignore Task.");
+      continue;
+    }
+
+    uint64_t target_start = target_zone->wp_;
+    if (zfile->IsSparse()) {
+      // For buffered write, ZenFS use inlined metadata for extents and each
+      // extent has a SPARSE_HEADER_SIZE.
+      target_start = target_zone->wp_ + ZoneFile::SPARSE_HEADER_SIZE;
+      zfile->MigrateData(ext->start_ - ZoneFile::SPARSE_HEADER_SIZE,
+                         ext->length_ + ZoneFile::SPARSE_HEADER_SIZE,
+                         target_zone);
+    } else {
+      zfile->MigrateData(ext->start_, ext->length_, target_zone);
+    }
+
+    // Check again if the file still exist
+    if (GetFileInternal(fname) == nullptr) {
+      Info(logger_, "Migrate file not exist anymore.");
+      zbd_->ReleaseMigrateZone(target_zone);
+      break;
+    }
+
+    // Update zone stats
+    ext->WriteLock();
+    ext->start_ = target_start;
+    ext->zone_ = target_zone;
+    ext->zone_->used_capacity_ += ext->length_;
+    ext->WriteUnlock();
+
+    old_zone->used_capacity_ -= ext->length_;
+
+    zbd_->ReleaseMigrateZone(target_zone);
+  }
+
+  /*
+  std::string record;
+  EncodeFileDeletionTo(zfile, &record);
+  s = PersistRecord(record);
+  if (!s.ok()) {
+    Info(logger_, "Migration: Delete old file failed!");
+    return s;
+  }
+  */
+
+  // Add a file replace record
+  zfile->MetadataUnsynced();
+  SyncFileMetadata(zfile);
```

Comment:
Hmm, this should be SyncFileMetadata(zfile, true), right?

Reply:
done